Ensembles

orionw opened the pull request:
This PR is after the Validation PR
- closes #3 

Features:
- Generates ensembles with a arbitrary number of sub-pipelines, with options for random and same-model predictors, random and same-model preprocessors, as well as specific models and model order.
- Test cases for generating ensembles and error checking
- Test cases for executing ensembles and simulating the whole process

We still need to decide how we want to use these: should I add a command line option?  Should we just do it manually?

orionw committed:
progress until now

orionw committed:
Merge branch 'develop' of https://github.com/byu-dml/d3m-experimenter into ensembles

orionw committed:
merged in docker updates

orionw committed:
working dev so far

orionw committed:
working to ensemble. Needs clarification

orionw committed:
merged validate branch in

orionw committed:
Fixed imports

orionw committed:
Imports, validation fix

orionw committed:
time changes

orionw committed:
Ensembles working

orionw committed:
Documentation, Error Checking

orionw committed:
Tests for Ensembling

They all pass! Tested execution and generation.

orionw committed:
Cleaning, refactor for PR

bjschoenfeld committed:
Merge branch 'develop' into ensembles

bjschoenfeld committed:
Merge branch 'develop' into ensembles

bjschoenfeld commented on experimenter/experimenter.py:
```diff
@@ -64,14 +65,21 @@ def __init__(self, datasets_dir: str, volumes_dir: str, input_problem_directory=
                 print('Exporting pipelines to mongodb...')
                 self.output_pipelines_to_mongodb()
+            elif location == "test":
+                pass
```
This does not seem like good practice to have test related code in real code. You can set location to a specific place and then clean it up after the test is over.

orionw replied:
The problem is that one of the test cases in test_generate_pipelines was creating a pipeline and then submitting it to mongodb.  I wanted to have an option to generate pipelines but not store them anywhere.  Perhaps a better name would be "NoStore" or something like that.  We could pass it a location for the test, but I thought that case deserved it's own option.  I'm fine either way though.

bjschoenfeld replied:
> The problem is that one of the test cases in test_generate_pipelines was creating a pipeline and then submitting it to mongodb.

Can we write the pipeline to a dummy location and then delete it later?

orionw replied:
Sure that works!  I just thought maybe we wanted an option to generate but not store right away.  I can't think of many uses offhand, so we can definitely go that route

bjschoenfeld replied:
Ideally (I think) we would not be writing to any location in the constructor. Rather, we would have methods to write to database or to write to file.

orionw replied:
Yes, I agree.  Do you want me to refactor it for this PR?

bjschoenfeld replied:
Let's not do the major refactoring for this PR. Let's just write the files to /dev/null for tests

orionw replied:
Do I need to change anything in this PR for this?  I'm trying to get this ready so we can merge it in

bjschoenfeld replied:
Yes, let's remove `location == "test"`. Us the path "/dev/null" in the test code when you don't want it to go anywhere.

orionw committed:
Ready to launch ensembles, find errors

Drop timeout, remove datasets that are too long, deterministic
ensembles, visualization

orionw committed:
Merge branch 'ensembles' of https://github.com/byu-dml/d3m-experimenter into ensembles

bjschoenfeld commented on experimenter/constants.py:
```diff
@@ -96,6 +94,50 @@
+    "uu2_gp_hyperparameter_estimation_v2",  # has extra data
+    "uu2_gp_hyperparameter_estimation",  # has extra data
+    "57_hypothyroid",  # always NAN's out
+
```
do we need all these blank lines?

orionw replied:
Nope. Deleting them

orionw committed:
Add regex for getting pipelines

find the problem type

orionw committed:
Add script to grab D3M documents

orionw committed:
Remove whitespace

bjschoenfeld commented on docker-compose.yml:
```diff
@@ -12,6 +12,7 @@ services:
       - ${VOLUMES}:/volumes
       - ${DATASETS}:/datasets
       - ${EXPERIMENTER}:/d3m-experimenter
+      - ${D3M}:/d3m
```
why is this needed? it is installed in the base image

orionw replied:
Yeah I think I was tired of adding it into all my branch work and then just decided to leave it.  I'll take it out

orionw committed:
Add failure type, document check prints, faster check

orionw committed:
Remove d3m from docker-compose

bjschoenfeld commented on tests/test_pipeline_generator.py:
```diff
@@ -20,7 +20,7 @@ def setUp(self):
             self.datasets_dir, volumes_dir,
             self.seed_problem_directory, models, preprocessors,
-            generate_problems=True
+            generate_problems=True, location="test"
```
Let's remove this, as above.

orionw replied:
It is removed!

Scowley4 committed:
Remove test code

epeters3 committed:
Add development DB, Dockerfile addition, and minor README fixes
